src/analysis/lattices/stack.h

```diff
@@ -74,193 +74,115 @@ namespace wasm::analysis {
-  // element starting from the top.
-  //
-  // If the left stack is higher, and left top >= right top, then we say
-  // that left stack > right stack. If the left stack is lower and the left top
-  // >= right top or if the left stack is higher and the right top > left top or
-  // they are unrelated, then there is no relation. Same applies for the reverse
-  // relationship.
-  LatticeComparison compare(const Element& left,
-                            const Element& right) const noexcept {
-    // Handle cases where there are top elements.
-    if (left.isTop()) {
-      if (right.isTop()) {
-        return LatticeComparison::EQUAL;
-      } else {
-        return LatticeComparison::GREATER;
+  LatticeComparison compare(const Element& a, const Element& b) const noexcept {
+    auto result = EQUAL;
+    // Iterate starting from the end of the vectors to match up the tops of
+    // stacks with different sizes.
+    for (auto itA = a.rbegin(), itB = b.rbegin();
+         itA != a.rend() || itB != b.rend();
+         ++itA, ++itB) {
+      if (itA == a.rend()) {
+        // The rest of A's elements are bottom, but B has a non-bottom element
+        // because of our invariant that the base of the materialized stack must
+        // not be bottom.
+        return LESS;
       }
-    } else if (right.isTop()) {
-      return LatticeComparison::LESS;
-    }
-
-    bool hasLess = false;
-    bool hasGreater = false;
-
-    // Check the tops of both stacks which match (i.e. are within the heights
-    // of both stacks). If there is a pair which is not related, the stacks
-    // cannot be related.
-    for (auto leftIt = left.stackValue->crbegin(),
-              rightIt = right.stackValue->crbegin();
-         leftIt != left.stackValue->crend() &&
-         rightIt != right.stackValue->crend();
-         ++leftIt, ++rightIt) {
-      LatticeComparison currComparison = lattice.compare(*leftIt, *rightIt);
-      switch (currComparison) {
-        case LatticeComparison::NO_RELATION:
-          return LatticeComparison::NO_RELATION;
-        case LatticeComparison::LESS:
-          hasLess = true;
-          break;
-        case LatticeComparison::GREATER:
-          hasGreater = true;
-          break;
-        default:
-          break;
+      if (itB == b.rend()) {
+        // The rest of B's elements are bottom, but A has a non-bottom element.
+        return GREATER;
+      }
+      switch (lattice.compare(*itA, *itB)) {
+        case NO_RELATION:
+          return NO_RELATION;
+        case EQUAL:
+          continue;
+        case LESS:
+          if (result == GREATER) {
+            // Cannot be both less and greater.
+            return NO_RELATION;
+          }
+          result = LESS;
+          continue;
+        case GREATER:
+          if (result == LESS) {
+            // Cannot be both greater and less.
+            return NO_RELATION;
+          }
+          result = GREATER;
+          continue;
```

Comment:
This switch sure looks familiar now... :thinking:

Reply:
Maaaaybe it makes sense to try to share the switch logic. Note that everything surrounding the switch is slightly different in every case, though. (And in the tuple case, the switch contains recursive calls instead of `continue`s.)